File: python/open_guardrail/guards/eu_ai_risk_classification.py
"""Auto-classify AI output by EU AI Act risk level."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_UNACCEPTABLE: list[re.Pattern[str]] = [
    re.compile(r"\bcitizen\s+score\b", re.IGNORECASE),
    re.compile(r"\bsocial\s+credit\b", re.IGNORECASE),
    re.compile(r"\bsocial\s+scoring\b", re.IGNORECASE),
    re.compile(
        r"\bwithout\s+your\s+knowledge\b", re.IGNORECASE
    ),
    re.compile(r"\bsubconsciously\b", re.IGNORECASE),
    re.compile(
        r"\bsubliminal\s+manipulation\b", re.IGNORECASE
    ),
    re.compile(
        r"\btargeting\s+children\b", re.IGNORECASE
    ),
    re.compile(
        r"\bexploiting\s+elderly\b", re.IGNORECASE
    ),
    re.compile(
        r"\bexploit(?:ing)?\s+vulnerabilit",
        re.IGNORECASE,
    ),
]

_HIGH_RISK: list[re.Pattern[str]] = [
    re.compile(
        r"\bfacial\s+recognition\b", re.IGNORECASE
    ),
    re.compile(r"\bfingerprint\b", re.IGNORECASE),
    re.compile(r"\bbiometric\b", re.IGNORECASE),
    re.compile(
        r"\bhiring\s+decision\b", re.IGNORECASE
    ),
    re.compile(
        r"\bcandidate\s+evaluation\b", re.IGNORECASE
    ),
    re.compile(
        r"\bemployment\s+screening\b", re.IGNORECASE
    ),
    re.compile(
        r"\bcredit\s+scor(?:e|ing)\b", re.IGNORECASE
    ),
    re.compile(
        r"\bloan\s+approval\b", re.IGNORECASE
    ),
    re.compile(
        r"\bcreditworthiness\b", re.IGNORECASE
    ),
    re.compile(
        r"\bcriminal\s+prediction\b", re.IGNORECASE
    ),
    re.compile(r"\brecidivism\b", re.IGNORECASE),
    re.compile(r"\bsurveillance\b", re.IGNORECASE),
    re.compile(
        r"\bstudent\s+assessment\b", re.IGNORECASE
    ),
    re.compile(
        r"\badmission\s+decision\b", re.IGNORECASE
    ),
    re.compile(r"\bgrading\b", re.IGNORECASE),
]

_LIMITED: list[re.Pattern[str]] = [
    re.compile(r"\bchatbot\b", re.IGNORECASE),
    re.compile(
        r"\bemotion\s+recognition\b", re.IGNORECASE
    ),
    re.compile(r"\bdeepfake\b", re.IGNORECASE),
]

_LEVEL_ORDER = {
    "minimal": 0,
    "limited": 1,
    "high": 2,
    "unacceptable": 3,
}


def _classify(text: str) -> str:
    if any(p.search(text) for p in _UNACCEPTABLE):
        return "unacceptable"
    if any(p.search(text) for p in _HIGH_RISK):
        return "high"
    if any(p.search(text) for p in _LIMITED):
        return "limited"
    return "minimal"


class _EuAiRiskClassification:
    def __init__(
        self,
        *,
        action: str = "warn",
        min_risk_level: str = "high",
    ) -> None:
        self.name = "eu-ai-risk-classification"
        self.action = action
        self._min_level = min_risk_level

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        level = _classify(text)
        triggered = (
            _LEVEL_ORDER[level]
            >= _LEVEL_ORDER[self._min_level]
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="eu-ai-risk-classification",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=(
                _LEVEL_ORDER[level] / 3
                if triggered
                else 0.0
            ),
            message=(
                f"Risk level: {level}"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details={"risk_level": level},
        )


def eu_ai_risk_classification(
    *, action: str = "warn", min_risk_level: str = "high"
) -> _EuAiRiskClassification:
    return _EuAiRiskClassification(
        action=action, min_risk_level=min_risk_level
    )

File: packages/guards/src/eu-ai-risk-classification.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

type RiskLevel = 'unacceptable' | 'high' | 'limited' | 'minimal';

interface EuAiRiskClassificationOptions {
  action: 'block' | 'warn';
  minRiskLevel?: RiskLevel;
}

const UNACCEPTABLE: RegExp[] = [
  /\bcitizen\s+score\b/i,
  /\bsocial\s+credit\b/i,
  /\bsocial\s+scoring\b/i,
  /\bwithout\s+your\s+knowledge\b/i,
  /\bsubconsciously\b/i,
  /\bsubliminal\s+manipulation\b/i,
  /\btargeting\s+children\b/i,
  /\bexploiting\s+elderly\b/i,
  /\bexploit(?:ing)?\s+vulnerabilit/i,
];

const HIGH_RISK: RegExp[] = [
  /\bfacial\s+recognition\b/i,
  /\bfingerprint\b/i,
  /\bbiometric\b/i,
  /\bhiring\s+decision\b/i,
  /\bcandidate\s+evaluation\b/i,
  /\bemployment\s+screening\b/i,
  /\bcredit\s+scor(?:e|ing)\b/i,
  /\bloan\s+approval\b/i,
  /\bcreditworthiness\b/i,
  /\bcriminal\s+prediction\b/i,
  /\brecidivism\b/i,
  /\bsurveillance\b/i,
  /\bstudent\s+assessment\b/i,
  /\badmission\s+decision\b/i,
  /\bgrading\b/i,
];

const LIMITED: RegExp[] = [
  /\bchatbot\b/i,
  /\bemotion\s+recognition\b/i,
  /\bdeepfake\b/i,
];

const LEVEL_ORDER: Record<RiskLevel, number> = {
  minimal: 0,
  limited: 1,
  high: 2,
  unacceptable: 3,
};

function classify(text: string): RiskLevel {
  if (UNACCEPTABLE.some((p) => p.test(text))) return 'unacceptable';
  if (HIGH_RISK.some((p) => p.test(text))) return 'high';
  if (LIMITED.some((p) => p.test(text))) return 'limited';
  return 'minimal';
}

export function euAiRiskClassification(options: EuAiRiskClassificationOptions): Guard {
  const minLevel = options.minRiskLevel ?? 'high';

  return {
    name: 'eu-ai-risk-classification',
    version: '0.1.0',
    description: 'Auto-classify AI output by EU AI Act risk level',
    category: 'compliance',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const level = classify(text);
      const triggered = LEVEL_ORDER[level] >= LEVEL_ORDER[minLevel];

      return {
        guardName: 'eu-ai-risk-classification',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score: triggered ? LEVEL_ORDER[level] / 3 : 0,
        latencyMs: Math.round(performance.now() - start),
        details: { riskLevel: level },
      };
    },
  };
}
